source/core/slang-memory-arena.h

```diff
@@ -0,0 +1,393 @@
+        If the arena has been previously initialized will free and deallocate all memory */
+    void init(size_t blockSize, size_t blockAlignment = kMinAlignment);
+
+        /** Allocate some memory of at least size bytes without having any specific alignment.
+
+         Can be used for slightly faster *aligned* allocations if caveats in class description are met. The
+         Unaligned, means the method will not enforce alignment - but a client call to allocateUnaligned can control
+         subsequent allocations alignments via it's size.
+
+         @param size The size of the allocation requested (in bytes and must be > 0).
+         @return The allocation. Can be nullptr if backing allocator was not able to request required memory */
+    void* allocate(size_t sizeInBytes);
+
+        /** Allocate some aligned memory of at least size bytes 
+         @param size Size of allocation wanted (must be > 0).
+         @param alignment Alignment of allocation - must be a power of 2.
+         @return The allocation (or nullptr if unable to allocate). Will be at least 'alignment' alignment or better. */
+    void* allocateAligned(size_t sizeInBytes, size_t alignment);
+
+        /** Allocate some aligned memory of at least size bytes 
+        @param size Size of allocation wanted (must be > 0).
+        @param alignment Alignment of allocation - must be a power of 2.
+        @return The allocation (or nullptr if unable to allocate). Will be at least 'alignment' alignment or better. */
+    void* allocateUnaligned(size_t sizeInBytes);
+
+        /** Allocates a null terminated string.
+        @param str A null-terminated string
+        @return A copy of the string held on the arena */
+    const char* allocateString(const char* str);
+
+        /** Allocates a null terminated string.     
+         @param chars Pointer to first character
+         @param charCount The amount of characters NOT including terminating 0.
+         @return A copy of the string held on the arena. */
+    const char* allocateString(const char* chars, size_t numChars);
+
+        /// Allocate an element of the specified type. Note: Constructor for type is not executed.
+    template <typename T>
+    T* allocate();
+
+        /// Allocate an array of a specified type. NOTE Constructor of T is NOT executed.
+    template <typename T>
+    T* allocateArray(size_t size);
+
+        /// Allocate an array of a specified type, and copy array passed into it.
+    template <typename T>
+    T* allocateAndCopyArray(const T* src, size_t size);  
+
+        /// Allocate an array of a specified type, and zero it.
+    template <typename T>
+    T* allocateAndZeroArray(size_t size);
+
+        /// Deallocate the last allocation. If data is not from the last allocation then the behavior is undefined.
+    void deallocateLast(void* data);
+
+        /// Deallocate this allocation and all remaining after it.
+    void deallocateAllFrom(void* dataStart);
```

Comment:
These are fair points. Because of some issues I have a new implementation that removes the ability to rewind. This is mainly because doing so allows better memory usage.